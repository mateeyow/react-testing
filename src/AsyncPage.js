import React, { useEffect, useState } from 'react';
import fetch from 'axios';

const AsyncPage = () => {
  const [data, setData] = useState([]);
  const [isError, setError] = useState(false);
  const [isLoading, setLoading] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      setError(false);
      setLoading(true);
      try {
        const response = await fetch.get(
          'https://jsonplaceholder.typicode.com/posts?_limit=10',
        );

        if (response) {
          const { data: posts } = response;
          setData(posts);
        }

        setLoading(false);
      } catch (err) {
        setError(true);
        setLoading(false);
      }
    };

    fetchData();
  }, []);

  if (isError) {
    return <div>Something went wrong</div>
  }

  if (isLoading) {
    return <div>Loading...</div>;
  }

  if (!data.length) {
    return <div>No data found</div>;
  }

  return (
    <div>
      {data.map(item => (
        <div key={item.id}>
          <h2>{item.title}</h2>
          <p>{item.body}</p>
        </div>
      ))}
    </div>
  );
};

export default AsyncPage;
