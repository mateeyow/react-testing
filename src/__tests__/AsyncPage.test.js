import React from 'react';
import { render, wait } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import * as fetch from 'axios';
import AsyncPage from '../AsyncPage';

jest.mock('axios');

describe('<AsyncPage /> tests', () => {
  afterEach(() => {
    fetch.mockRestore()
  })

  it('should render the page by default', async () => {
    const text = 'No data found';
    fetch.get.mockResolvedValue({ data: [] });

    const { container, getByText } = render(<AsyncPage />);

    await wait()

    expect(container.firstChild).toMatchSnapshot();
    expect(getByText(text)).toBeInTheDocument();
  });

  it('should fetch data successfully', async () => {
    const posts = [{ id: 1, title: 'sample title', body: 'sample body' }];
    fetch.get.mockResolvedValue({ data: posts });

    const { getByText, container } = render(<AsyncPage />);

    expect(getByText('Loading...')).toBeInTheDocument();

    await wait()
    expect(getByText('sample title')).toBeInTheDocument();
    expect(container).toMatchSnapshot()
  });

  it('should show an error message', async () => {
    const text = 'Something went wrong'
    fetch.get.mockImplementation(() => Promise.reject('Some error'))

    const { getByText, container } = render(<AsyncPage />);

    await wait()

    expect(getByText(text)).toBeInTheDocument(0)
    expect(container).toMatchSnapshot()
  })
});
