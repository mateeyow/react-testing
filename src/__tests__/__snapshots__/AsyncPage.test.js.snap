// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncPage /> tests should fetch data successfully 1`] = `
<div>
  <div>
    <div>
      <h2>
        sample title
      </h2>
      <p>
        sample body
      </p>
    </div>
  </div>
</div>
`;

exports[`<AsyncPage /> tests should render the page by default 1`] = `
<div>
  No data found
</div>
`;

exports[`<AsyncPage /> tests should show an error message 1`] = `
<div>
  <div>
    Something went wrong
  </div>
</div>
`;
